nodedb.go

```diff
@@ -956,6 +815,55 @@ func (ndb *nodeDB) decrVersionReaders(version int64) {
 	}
 }
 
+// traverseOrphans traverses orphans which removed by the updates of the version (n+1).
+func (ndb *nodeDB) traverseOrphans(version int64, fn func(*Node) error) error {
+	cRoot, err := ndb.getRoot(version + 1)
+	if err != nil {
+		return err
+	}
+
+	curIter, err := NewNodeIterator(cRoot, ndb)
+	if err != nil {
+		return err
+	}
+
+	pRoot, err := ndb.getRoot(version)
+	if err != nil {
+		return err
+	}
+	prevIter, err := NewNodeIterator(pRoot, ndb)
+	if err != nil {
+		return err
+	}
+
+	var orgNode *Node
+	for prevIter.Valid() {
+		for orgNode == nil && curIter.Valid() {
+			node := curIter.GetNode()
+			if node.version <= version {
+				curIter.Next(true)
+				orgNode = node
+			} else {
+				curIter.Next(false)
+			}
+		}
+		pNode := prevIter.GetNode()
+
+		if orgNode != nil && bytes.Equal(pNode.hash, orgNode.hash) {
```

Comment:
Are you sure preorder traversal is correct for this, I'm still trying to reason about the correctness of your algorithm, it seems we need to inorder traversal here which is ordered by node keys.

Reply:
But we do need preorder to try to skip the subtree early on, so we need to check twice for a branch node both preorder and inorder.

Reply:
pre-order or in-order are not important here, right? Because we only care about the order of leaves. I don't like keep orgNodes or orphans in memory, those amounts may be enormous.

Reply:
No, orgNode could be branch nodes